service/cziti/dns-proxy.go

```diff
@@ -341,69 +344,117 @@ outer:
-						pr.req.Question[0].Name,
-						proxy.RemoteAddr(),
-						err)
-				} else {
-					log.Tracef("Proxied request sent to %v from ipv%d listener", proxy.RemoteAddr(), pr.ipVer)
+	dnsReqChannel := make(chan struct{})
+	ticker := time.NewTicker(5 * time.Second)
+	defer func() {
+		close(dnsReqChannel)
+		log.Tracef("Exiting dns request handling routines.")
+	}()
+	go func() {
+	CleanUpIteration:
+		for {
+			select {
+			case <-ticker.C:
+				if len(reqs) == 0 {
+					continue CleanUpIteration
+				}
+				// cleanup requests we didn't get answers for
+				now := time.Now()
+				dnsReqMutex.Lock()
+				for k, r := range reqs {
+					if now.After(r.exp) {
+						log.Debugf("a DNS request has expired - enable trace logging and reproduce this issue for more information")
+						log.Tracef("expired DNS req: %s %s", dns.Type(r.req.Question[0].Qtype), r.req.Question[0].Name)
+
+						delete(reqs, k)
+					}
 				}
+				dnsReqMutex.Unlock()
+			case <-dnsReqChannel:
+				ticker.Stop()
+				return
 			}
+		}
+	}()
 
-		case rep := <-respChan:
-			reply := dns.Msg{}
-			if err := reply.Unpack(rep); err == nil {
-				id := (uint32(reply.Id) << 16) | uint32(reply.Question[0].Qtype)
-				req, found := reqs[id]
-				if found {
-					delete(reqs, id)
-					log.Tracef("proxy resolved request for %v id:%d", reply.Question[0].Name, reply.Id)
-					n, oobn, err := req.s.WriteMsgUDP(rep, nil, req.peer)
-					if err != nil {
-						log.Errorf("an error has occurred while trying to write a udp message. n:%d, oobn:%d, err:%v", n, oobn, err)
+	go func() {
+		for {
+			select {
+			case rep := <-respChan:
+				reply := dns.Msg{}
+				if err := reply.Unpack(rep); err == nil {
+					id := (uint32(reply.Id) << 16) | uint32(reply.Question[0].Qtype)
+					dnsReqMutex.Lock()
+					req, found := reqs[id]
+					if found {
+						delete(reqs, id)
+						dnsReqMutex.Unlock()
+						log.Tracef("proxy resolved request for %v id:%d", reply.Question[0].Name, reply.Id)
+						n, oobn, err := req.s.WriteMsgUDP(rep, nil, req.peer)
+						if err != nil {
+							log.Errorf("an error has occurred while trying to write a udp message. n:%d, oobn:%d, err:%v", n, oobn, err)
+						}
+					} else {
+						dnsReqMutex.Unlock()
+						// keep this log but leave commented out. When two listeners are enabled (ipv4/v6) this msg will
+						// just mean some other request was processed successfully and removed the entry from the map
+						// log.Tracef("matching request was not found for id:%d. %s %s", reply.Id, dns.Type(reply.Question[0].Qtype), reply.Question[0].Name)
 					}
-				} else {
-					// keep this log but leave commented out. When two listeners are enabled (ipv4/v6) this msg will
-					// just mean some other request was processed successfully and removed the entry from the map
-					// log.Tracef("matching request was not found for id:%d. %s %s", reply.Id, dns.Type(reply.Question[0].Qtype), reply.Question[0].Name)
 				}
+			case <-dnsReqChannel:
+				return
```

Comment:
it'd be good to have a log here to make sure we know when this goroutine exits

Reply:
Added trace logs